### PR TITLE
Update Value of TwoDigitYearMax to 2049

### LINKED ISSUE #75148: Consider increase to TwoDigitYearMax
TwoDigitYearMax is the century assumption we use when parsing a date string with a two digit year. Currently all our calendars seem to use 29, ie., 29 implies 2029 but 30 implies 1930. [This ](https://docs.microsoft.com/en-us/office/troubleshoot/excel/two-digit-year-numbers#using-the-regional-settings-in-control-panel) suggests back in Windows 98, Windows chose 29 and that's where we got 29 from. At least back to 2.0, we seem to have been defaulting to 29 if we couldn't read the system setting. We no longer read Windows NLS settings by default, and ICU does not have such data, so the default of 29 applies almost everywhere, unless overridden in code.

so eg -- 
```c#
        var c = CultureInfo.GetCultureInfo("en-US").Calendar;
        Console.WriteLine($"{c.GetType()} calendar has TwoDigitYearMax of {c.TwoDigitYearMax}");
        
        var dt = DateOnly.Parse("10/10/28");
        Console.WriteLine(dt);        
        
        dt = DateOnly.Parse("10/10/35");
        Console.WriteLine(dt);
```
gives
```
System.Globalization.GregorianCalendar calendar has TwoDigitYearMax of 2029
10/10/2028
10/10/1935
```

While Excel still uses 29, Windows now has moved to 49 for at least some calendars. Currently on my Windows box, with the en-US culture and Gregorian calendar, if I run with DOTNET_SYSTEM_GLOBALIZATION_USENLS=1 then I get `System.Globalization.GregorianCalendar calendar has TwoDigitYearMax of 2049`. Windows allows you to change this setting although it's deep in the UI:

image

Now that we're controlling this value by default on all OS, the value has become much more important than on .NET Framework. We may choose to make a similar decision as we are approaching 2030. The closer we get to 2030 the more likely it will be that if I type "1/1/30" I mean 2030.

Options
1. Leave as-is. Document that users should explicitly increase TwoDigitYearMax if they parse 2-digit years and the data has a different assumption.
2. Leave as-is, but on Windows when using ICU, call specially into NLS for the two digit cutoff, since ICU has no value. This makes Windows inconsistent with other OS, by default.
3. Change to something larger that gives wiggle room, eg 49 (there appears to be no consensus, see below) possibly with a global config switch to revert to 29, similar to ICU/NLS. Doc as a breaking change.

Prior art:

Python uses a [cutoff of 68](https://docs.python.org/3/library/time.html)
Javascript uses a [cutoff of 49](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getFullYear)
Java seems to be [99](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html)
Golang [seems to be 68](https://pkg.go.dev/time#Parse)
Nodatime [seems to be 30](https://nodatime.org/3.0.x/userguide/localdate-patterns)
Excel uses 29. It respects Windows settings on Windows, but only for user input. 
Rust uses 69 in at least one instance - I experimented naively with something like `NaiveDate::parse_from_str("70-09-05", "%y-%m-%d")` 
Perl - not sure, I found [one reference to 50](http://marms.sourceforge.net/perl/modules/Year.html), another package (DateTime::Format::Flexible) uses 69
RFC3339 [just says](https://www.rfc-editor.org/rfc/rfc3339#page-4) the internet should always use 4 digit years. 
Perhaps there are other relevant standards/RFC's ? 

I don't have Linux/macOS handy. Do they have similar system settings, possibly user modifiable, and separate from ICU? If so, possibly we should read those values on all OS. This would however further increase the possibility of introducing a bug when, eg., developing on macOS and deploying on Linux.

### PR DESCRIPTION
**Pull request by cdbullard**
Fix #75148 

Update the century assumption cutoff year for determining whether a year is referring to a 19XX date or a 20XX date from 2029 to 2049.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs [hunk 1]
```diff
@@ -40,7 +40,7 @@ internal sealed partial class CalendarData
         internal string[] saLeapYearMonthNames = null!; // Multiple strings for the month names in a leap year.
 
         // Integers at end to make marshaller happier
-        internal int iTwoDigitYearMax = 2029; // Max 2 digit year (for Y2K bug data entry)
+        internal int iTwoDigitYearMax = 2049; // Max 2 digit year (for Y2K bug data entry)
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:43]: I'm not sure what `(for Y2K bug data entry)` means but it might be a very obsolete comment at this point. I'd remove it

File: src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs [hunk 2]
```diff
@@ -646,7 +646,7 @@ public override bool IsLeapYear(int year, int era)
             return InternalIsLeapYear(year);
         }
 
-        private const int DefaultGregorianTwoDigitYearMax = 2029;
+        private const int DefaultGregorianTwoDigitYearMax = 2049;
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:649]: incidentally, I am not sure why this
https://github.com/dotnet/runtime/blob/43ebf1a159c0af6ea8c4364d1e9d881972561914/src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs#L657
and this are different
https://github.com/dotnet/runtime/blob/43ebf1a159c0af6ea8c4364d1e9d881972561914/src/libraries/System.Private.CoreLib/src/System/Globalization/GregorianCalendar.cs#L417

**Comment by danmoseley** [comment]
Tests are all under System.Globalization.Calendars. Search for "twodigityear" eg
https://github.com/dotnet/runtime/blob/dd6696a29c77c3256d35ef25970fba7c09440aca/src/libraries/System.Globalization.Calendars/tests/GregorianCalendar/GregorianCalendarTwoDigitYearMax.cs#L25

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs (hunk 1 above)
Comment (cdbullard) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs]: Sure thing, this has been removed

File: src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs (hunk 2 above)
Comment (cdbullard) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:649]: In the EastAsianLunisolarCalendar file, I've made the call to GetSystemTwoDigitYearSetting match how it's handled in GregorianCalendar. Let me know if it should be the other way around

File: src/libraries/System.Globalization.Calendars/System.Globalization.Calendars.sln [hunk 3]
```diff
@@ -1,4 +1,8 @@
-﻿Microsoft Visual Studio Solution File, Format Version 12.00
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio Version 17
+VisualStudioVersion = 17.3.32922.545
```
Comment (danmoseley) [inline src/libraries/System.Globalization.Calendars/System.Globalization.Calendars.sln:4]: Please remove sln edits from this change...I believe we generate these.

Comment (cdbullard) [reply src/libraries/System.Globalization.Calendars/System.Globalization.Calendars.sln]: This has been corrected

**Comment by cdbullard** [comment]
Interestingly when I try to change this to 2049 and 2069, the unit test fails: https://github.com/dotnet/runtime/blob/dd6696a29c77c3256d35ef25970fba7c09440aca/src/libraries/System.Globalization.Calendars/tests/GregorianCalendar/GregorianCalendarTwoDigitYearMax.cs#L25

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:43]: I still see it here and below :)

Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:43]: @elinor-fung do you know what "// Integers at end to make marshaller happier" means? just curious

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:43]: `for Y2K bug data entry` was a helper comment telling how important the two digit year max for Y2K problem when mostly two digit years was used than four digit years.

Comment (elinor-fung) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:43]: I don't know what that is referring to - marshalling should not require integers at the end... It also doesn't look like `CalendarData` is marshalled in any way now (I assume it was once upon a time).

In any case, I think the comment should be removed.

**Comment by danmoseley** [comment]
some extra context here @cdbullard, although you probably know
* on non Windows, we use the ICU library for globalization settings. we don't read two digit year value from ICU, apparently it doesn't exist in ICU, so we use the hard coded default you changed.
* on Windows we also use ICU, so again we have to use the hard coded default. however, you can switch your app to NLS mode if you want, and in this mode is needs to read the Windows setting you can change in the Window UI. This is the behavior today and should continue.

This has info on how to switch to NLS mode. Before your change, you should notice that in NLS mode, when you change the Windows two digit year value in the Windows UI, your C# code will see the change. That's the behavior that needs to stay.
https://learn.microsoft.com/en-US/dotnet/core/runtime-config/globalization

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs (hunk 1 above)
Comment (cdbullard) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs]: > I still see it here and below :)

Whoops--I mixed up the comments and thought this was talking about the .sln file changes. The Y2K note on line 43 has been removed

File: src/libraries/System.Private.CoreLib/src/System/Globalization/Calendar.cs [hunk 4]
```diff
@@ -711,7 +712,7 @@ internal static long TimeToTicks(int hour, int minute, int second, int milliseco
 
         internal static int GetSystemTwoDigitYearSetting(CalendarId CalID, int defaultYearValue)
         {
-            int twoDigitYearMax = GlobalizationMode.UseNls ? CalendarData.NlsGetTwoDigitYearMax(CalID) : CalendarData.IcuGetTwoDigitYearMax();
+            int twoDigitYearMax = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? CalendarData.NlsGetTwoDigitYearMax(CalID) : CalendarData.IcuGetTwoDigitYearMax();
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/Calendar.cs:715]: Thanks @cdbullard for adding that. We usually avoid doing the OS checks inside the code. Instead we separate the code during compile time. 

For this change, please move the method `NlsGetTwoDigitYearMax` to the file `CalendarData.Windows.cs` and rename the method to something like `GetTwoDigitYearMax`. Then move the method `IcuGetTwoDigitYearMax` to the file `CalendarData.Unix.cs` and rename the method to the same name `GetTwoDigitYearMax` (taking same parameters as the one moved to Windows file). last, make `GetSystemTwoDigitYearSetting` just `CalendarData.GetTwoDigitYearMax`.

Comment (cdbullard) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/Calendar.cs:715]: Thanks for clarifying that for me. If I make the method `GetSystemTwoDigitYearSetting` like this:

```
internal static int GetSystemTwoDigitYearSetting(CalendarId CalID, int defaultYearValue)
{
    return CalendarData.GetTwoDigitYearMax(CalID);
}
```

should I also remove the `defaultYearValue` from the signature and from the places this method is being called?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/Calendar.cs:715]: No, please keep it as:

```C#
internal static int GetSystemTwoDigitYearSetting(CalendarId CalID, int defaultYearValue)
{
    int twoDigitYearMax = CalendarData.GetTwoDigitYearMax(CalID);
    return twoDigitYearMax >= 0 ? twoDigitYearMax : defaultYearValue;
}
```

Sorry I was not clear in that part :-)

Comment (cdbullard) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/Calendar.cs]: Thanks! I've just pushed a commit for this, let me know if any of it should be changed

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs [hunk 5]
```diff
@@ -12,5 +14,14 @@ internal sealed partial class CalendarData
...
 #pragma warning restore IDE0060
-     }
+
+        internal static int GetTwoDigitYearMax(CalendarId calendarId)
+        {
+            Debug.Assert(!GlobalizationMode.UseNls);
+
+            // There is no user override for this value on Linux or in ICU.
+            // So just return -1 to use the hard-coded defaults.
+            return -1;
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs:24]: >return -1; [](http://example.com/codeflow?start=12&length=10)

Could you please change this to:

```C#
return GlobalizationMode.Invariant ? Invariant.iTwoDigitYearMax : -1;
```

**Comment by tarekgh** [comment]
@cdbullard I added one more little comment, LGTM otherwise. Let's see if all tests pass and then we merge it.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs (hunk 5 above)
Comment (cdbullard) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs]: Sure thing, this has been corrected

**Comment by tarekgh** [comment]
@cdbullard 

```
src\libraries\System.Private.CoreLib\src\System\Globalization\CalendarData.Nls.cs(26,28): error CA1823: (NETCORE_ENGINEERING_TELEMETRY=Build) Unused field 'CAL_ITWODIGITYEARMAX'
```

please remove `CAL_ITWODIGITYEARMAX` from that file and ensure it is defined in the `CalendarData.Windows.cs` file.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs (hunk 5 above)
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Unix.cs]: please check there's a test that reads this in invariant mode (ie that would fail without the change above)

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs [hunk 6]
```diff
@@ -66,7 +65,7 @@ private static CalendarData CreateInvariant()
             invariant.sNativeName = "Gregorian Calendar";  // Calendar Name
 
             // Year
-            invariant.iTwoDigitYearMax = 2029; // Max 2 digit year (for Y2K bug data entry)
+            invariant.iTwoDigitYearMax = 2049; // Max 2 digit year (for Y2K bug data entry)
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:68]:
```suggestion
            invariant.iTwoDigitYearMax = 2049; // Max 2 digit year
```

File: src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:649]: I don't see that change.

**Comment by danmoseley** [comment]
did you merge away some of your tests? I only see one

File: src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs [hunk 7]
```diff
@@ -646,15 +646,15 @@ public override bool IsLeapYear(int year, int era)
...
 
         public override int TwoDigitYearMax
         {
             get
             {
                 if (_twoDigitYearMax == -1)
                 {
-                    _twoDigitYearMax = GetSystemTwoDigitYearSetting(BaseCalendarID, GetYear(new DateTime(DefaultGregorianTwoDigitYearMax, 1, 1)));
+                    _twoDigitYearMax = GetSystemTwoDigitYearSetting(ID, DefaultGregorianTwoDigitYearMax);
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:657]: Why did you change this? `GetYear` gets the year from the used calendar and not necessary to match the Gregorian year.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs (hunk 2 above)
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:649]: > incidentally, I am not sure why this

`GetYear` maps the Gregorian year to the calendar year. I don't think we need to change that.

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:649]: > In the EastAsianLunisolarCalendar file, I've made the call to GetSystemTwoDigitYearSetting match how it's handled in GregorianCalendar. Let me know if it should be the other way around

EastAsianLunisolarCalendar is abstract class which any user can subclass and introduce a new calendar. You never know what the calendar does. So we need to avoid any breaking with that.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs (hunk 7 above)
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:657]: that was my bad.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs (hunk 2 above)
Comment (cdbullard) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:649]: This change should now be reverted to the original code

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:649]: why?

Comment (cdbullard) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:649]: Sorry, I meant I have reverted the change to the `GetSystemTwoDigitYearSetting` call to the original code per your request

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/EastAsianLunisolarCalendar.cs:649]: Got it :-) thanks!

**Review by tarekgh — APPROVED** [review]

**Comment by tarekgh** [comment]
Thanks @cdbullard for helping with this change!

**Comment by cdbullard** [comment]
My pleasure @tarekgh! Will keep an eye out for more :-)
